pallets/connectors-gateway/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+	use cfg_types::domain_address::{Domain, DomainAddress};
+	use codec::{EncodeLike, FullCodec};
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::OriginFor;
+	use sp_std::convert::TryInto;
+
+	use super::*;
+	use crate::weights::WeightInfo;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::origin]
+	pub type Origin = GatewayOrigin;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The origin type.
+		type RuntimeOrigin: Into<Result<GatewayOrigin, <Self as frame_system::Config>::RuntimeOrigin>>
+			+ From<GatewayOrigin>;
+
+		/// The event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The LocalOrigin ensures that some calls can only be performed from a
+		/// local context i.e. a different pallet.
+		type LocalOrigin: EnsureOrigin<
+			<Self as frame_system::Config>::RuntimeOrigin,
+			Success = DomainAddress,
+		>;
+
+		/// The AdminOrigin ensures that some calls can only be performed by
+		/// admins.
+		type AdminOrigin: EnsureOrigin<<Self as frame_system::Config>::RuntimeOrigin>;
+
+		/// The incoming and outgoing message type.
+		///
+		/// NOTE - this `Codec` trait is the Centrifuge trait for connectors
+		/// messages.
+		type Message: Codec;
+
+		/// The message router type that is stored for each domain.
+		type Router: DomainRouter<Sender = Self::AccountId, Message = Self::Message>
+			+ Clone
+			+ Debug
+			+ MaxEncodedLen
+			+ TypeInfo
+			+ FullCodec
+			+ EncodeLike
+			+ PartialEq;
+
+		/// The type that processes incoming messages.
+		type InboundQueue: InboundQueue<Sender = Domain, Message = Self::Message>;
+
+		type WeightInfo: WeightInfo;
+
+		/// Maximum number of connectors for a domain.
+		#[pallet::constant]
+		type MaxConnectorsPerDomain: Get<u32>;
+
+		/// Maximum size of an incoming message.
+		#[pallet::constant]
+		type MaxIncomingMessageSize: Get<u32>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The router for a given domain was set.
+		DomainRouterSet { domain: Domain, router: T::Router },
+
+		/// A connector was added to a domain.
+		ConnectorAdded { connector: DomainAddress },
+
+		/// A connector was removed from a domain.
+		ConnectorRemoved { connector: DomainAddress },
+	}
+
+	/// Storage for domain routers.
+	///
+	/// This can only be set by an admin.
+	#[pallet::storage]
+	pub(crate) type DomainRouters<T: Config> = StorageMap<_, Blake2_128Concat, Domain, T::Router>;
+
+	/// Storage that contains a limited number of whitelisted connectors for a
+	/// particular domain.
+	///
+	/// This can only be modified by an admin.
+	#[pallet::storage]
+	pub(crate) type ConnectorsAllowlist<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		Domain,
+		BoundedVec<DomainAddress, T::MaxConnectorsPerDomain>,
+		ValueQuery,
+	>;
```

Comment:
@mustermeiszer @cdamian what about making this a `StorageDoubleMap`?  By doing so, we don't have the `O(n)` expense of checking whether a connector address is whitelisted for a given `Domain` but rather we can make it `O(1)`.

If we still need the `MaxConnectorsPerDomain` max in that case, we can always check that the `ConnectorsAllowlist::get(domain).keys().len() <= MaxConnectorsPerDomain`.

Regarding details, we could make the `Value` of the `StorageDoubleMap` a simple unit (`()`):

``` rust
#[pallet::storage]
	pub(crate) type ConnectorsAllowlist<T: Config> = StorageDoubleMap<
		_,
		Blake2_128Concat,
		Domain,
    Blake2_128Concat,
		DomainAddress,
    ()
		ValueQuery,
	>;
```

Reply:
I'm cool with this, awesome suggestion!

Reply:
Done, thanks again!